### PR TITLE
Configurable Specifications

### PR DESCRIPTION
**Pull request by lestrrat**
In #7 we got a request to add `%L` as a specification to generate textual representation of milliseconds from the given time. This in itself is totally fine, except %L is not in POSIX, and implementations differ on the exact pattern to invoke this (e.g. `%f` or `%L`).

While #7 just decided that %L was the right choice, I'd like to take this opportunity and introduce
enough flexibility that we can create different flavors of Strftime in the future when the needs arises.

This PR takes the previously static list of specifications, and moves them in a map so that they are more easily configurable. For common specifications, an optional parameter in the various methods can dynamically alter the specification set so that one instance of Strftime can generate a completely different set of text representations if need be.

If in the future we have enough flavors to work with, such as "the pythin strftime" or "the ruby strftime", we could just wrap the objects as:

```
... = strftime.Python.Format(t)
... = strftime.Ruby.Format(t)
... = strftime.POSIX.Format(t) // current default
```

And users can more easily navigate through the different implementations.

### CODE REVIEW

**Comment by moznion** [comment]
This change looks so great, thank you so much!

But I'm just curious, does this implementation support a pattern with multiple characters (e.g. `%02d`)?
Yes, I know this could be a bit too tricky and off-topic of this pull-request, but I just wish it supported that (even if not so, absolutely I don't mind☺).

File: README.md [hunk 1]
```diff
@@ -88,6 +88,56 @@ Formats the time according to the pre-compiled pattern, and returns the result s
...
+
+For those instances, `strftime.Strftime` can be configured to use a custom set of
+specifications:
+
+```
+ds := strftime.NewSpecificationSet()
+ds.Set('L', ...) // provide implementation for `%L`
+
+// pass this new specification set to the strftime instance
+p, err := strftime.New(`%L`, strftime.WithSpecificationSet(ss))
```
Comment (moznion) [inline README.md]: Should `ss` be `ds`?

**Comment by lestrrat** [comment]
IIRC, variable widths were not part of the POSIX specification.

But, I mean... Unlike sprintf, we know that days are going to be 2 digits max, years 4 or above, etc, so in _general_ I don't see much need for them -- There's only a limited set of specifications that can benefit from variable width specifiers. However, milliseconds may be one of them, and there may be other exceptions that I have overlooked.

In any case, I think this PR clears the way to creating a separate strftime formatter that respects those

e.g.

```
... = strftime.New(`%06L`, strftime.WithSpecification('L', strftime.VariableWidthMillisecond())
```

**Comment by lestrrat** [comment]
修正してマージしようとしたところで strftime.Format (インスタンスを作らないほう）が 3 alloc -> 17 allocになってるのを発見したのでどうしようか考えているところ

**Comment by moznion** [comment]
Woo, that is tough to choose which to go... but my opinion is "fast is right" ⚡️

**Comment by lestrrat** [comment]
17どころか 2 allocs/iteration まで落としてやった。

**Comment by lestrrat** [comment]
おし、マージします

**Comment by moznion** [comment]
ありがとうございます！
Thank you so much 🎉

**Pull request merged**